File: Two_Pointers/056.Merge-Intervals/056.Merge-Intervals.ts
function merge(intervals: number[][]): number[][] {
    intervals.sort((a, b) => a[0] - b[0]);
    const ret = [];
    let left = intervals[0][0];
    let right = intervals[0][1];

    for (let i = 1; i < intervals.length; i++) {
        if (intervals[i][0] >= left && intervals[i][0] <= right) {
            left = Math.min(left, intervals[i][0]);
            right = Math.max(right, intervals[i][1]);
        } else {
            ret.push([left, right]);
            left = intervals[i][0];
            right = intervals[i][1];
        }
    }
    
    ret.push([left, right]);

    return ret;
};

File: Two_Pointers/056.Merge-Intervals/056.Merge-Intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key=lambda x: x[0])
        result = []

        start = intervals[0][0]
        end = intervals[0][1]
        
        for i in range(len(intervals)):
            if intervals[i][0] <= end:
                end = max(intervals[i][1], end)
            else:
                result.append([start, end])
                start = intervals[i][0]
                end = intervals[i][1]
        result.append([start, end])
        return result
